Poprawilam selektory , ale niestety dalej pobiera jeden dom . tyle ze…

anetapec opened the pull request:
Poprawilam selektory , ale niestety dalej pobiera jeden dom . ..tyle ze juz nie w nieskonczonosc....

anetapec committed:
Poprawilam selektory , ale niestety dalej pobiera jeden dom . tyle ze juz nie w nieskonczonosc....

anetapec committed:
dodanie nastepnej strony do scrapowania

bagaw1 committed:
add launch json

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
     start_urls = ['https://ogloszenia.trojmiasto.pl/nieruchomosci/dom/gdynia/']
```
W tym folderze jest z duplikowany plik tricity\requiments.txt. Ten plik powinien być tylko w głównym foldrze. Ten z tego folderu trzeba usunąć.

anetapec replied:
usuniete

bagaw1 commented on tricity/tricity/tricity.csv:
```diff
@@ -0,0 +1,31 @@
+area,number_of_rooms,price,price_per_meter,url
```
Ten plik trzeba usunąć. z gita. Usuń ten plik i dodaj tą zmianę usunięcia pliku do gita. Dodaj do pliku .gitignore w głównym folderze "*.csv" tak żeby git ignorował tego typy pliki w przyszłości.

bagaw1 commented on tricity/tricity2.csv:
```diff
@@ -0,0 +1,93 @@
+area,number_of_rooms,price,price_per_meter,url
```
Ten plik też trzeba usunąć z git a. Nie będziemy trzymać plików csv w git.

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
     start_urls = ['https://ogloszenia.trojmiasto.pl/nieruchomosci/dom/gdynia/']
 
```
Dodaj do pliku README.md w folderze głównym w jaki sposób odpalać tego spidera

bagaw1 replied:
Przykładowa zawartość pliku README.md
# hip-performance-tests


Create virtual environment
```bash
cd .
python -m venv .venv
```

Activate VENV
```bash
.\.venv\Scripts\Activate.ps1
```

Install from requirements
```bash
pip install -r requirements.txt
```

### Running headless instance
```bash
locust -f .\sc1-ext.py --headless --users 10 --spawn-rate 1  -H https://ppe.westeurope.api.hip.inter.ikea.net
```


```bash
locust -f .\sc1-ext.py --users 10 --spawn-rate 1 -H https://ppe.westeurope.api.hip.inter.ikea.net

```

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
-        houses = response.xpath("//div[@id = 'ogl__list__wrap']").get()
+        houses = response.xpath('//div[contains(@id,"ogl-")]').getall()
 
         for house in houses:
```
Zmień wszystkie XPATHy na CSSSelectors. Od teraz będziemy używać tylko i wyłącznie CSSSelectors

bagaw1 replied:
`house`  jest instancją obiektu która pobierasz do pętli `for`. Nie jest ani razu użyta. Dlatego też cały czas pobierasz dane z jednego elementu.

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
             house_item = HouseItem()
-            house_item['price'] = response.xpath("//p[@class = 'list__item__price__value']//text()").get()
-            house_item['price_per_meter'] = response.css('p.list__item__details__info.details--info--price::text').get()
+            house_item['price'] = response.css('p.list__item__price__value::text').get().replace('zł', '')
```
tutaj trzeba odnieść się do elementu który pobierasz do pętli 'for' czyli `house`,  więc targetowanie poprzez CSS powinno być 
house.css('p.list__item__price__value::text').get().replace('zł', '').
Nie wiem czy jak zmienisz element z `response` na `house` to twoj selector css dalej będzie działać. Możesz popróbować za pomocą albo Debugger'a w VSCode albo za pomocą `scrapy shell`.  Debugowanie to jest narzędzie którym posługujemy się gdy coś nie działa i nie wiemy czemu. Wtedy można sprawdzić linijkę po linijce podczas odpalania kodu.
https://docs.scrapy.org/en/latest/topics/shell.html

bagaw1 replied:
Żeby dostać wszystkie wyświetlane domy możesz np. użyć selectora css `'#ogl__list__wrap .list__item'`

bagaw1 committed:
dodaj launch.json

bagaw1 committed:
add scrapy crawl command

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
             house_item['url'] = response.css('a.list__item__content__title__name.link').attrib['href']
-            house_item['area'] = response.xpath("//p[@class='list__item__details__icons__element__desc']//text()").extract_first()
-            house_item['number_of_rooms'] = response.xpath('//ul/li[2]/div/p[2]//text()').extract_first()
+            house_item['area'] = response.xpath("//p[@class='list__item__details__icons__element__desc']//text()").get().replace('\n                                            ', '').replace('\n                                    ', '')
```
Jak chcesz usunąć białe znaki to trzeba je namierzyć za pomocą wyrażenia regularnego a dosłownej ilości białych znaków.
Tutaj link to poczytania i poćwiczenia.
https://regexone.com/lesson/whitespaces

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
-            house_item['area'] = response.xpath("//p[@class='list__item__details__icons__element__desc']//text()").extract_first()
-            house_item['number_of_rooms'] = response.xpath('//ul/li[2]/div/p[2]//text()').extract_first()
+            house_item['area'] = response.xpath("//p[@class='list__item__details__icons__element__desc']//text()").get().replace('\n                                            ', '').replace('\n                                    ', '')
+            house_item['number_of_rooms'] = response.xpath('//ul/li[2]/div/p[2]//text()').get().replace('\n                                            ', '').replace('\n                                    ', '')
```
Tutaj też regex na usunięcie buałych znaków i zastąp XPatha za pomocą selectora CSS

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
\ No newline at end of file
+            yield house_item
+
+        next_page = response.css('span.pages__item__active').get()
```
Do przełączania się na nową stronę to musisz pobrać wartość atrybutu HREF z elementu HTML <a href=">.  Tym elementem na stronie jest `>` przełącza na kolejną stronę. Tutaj jak to można zrobić z przykładem
https://stackoverflow.com/questions/21181628/get-href-using-css-selector-with-scrapy.

Jak będziesz miała tą wartość to będzie coś jak "?strona=1", to wtedy musisz to połączyć z URL strony, można to pobrać albo z array  start_urls (linijka nr. 7 ) który jest property klasy HousesspiderSpider, albo w debuggerze można sprawdzić czy obiekt `response` nie ma jakiejś informacji którą można użyć.

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -7,14 +7,22 @@ class HousesspiderSpider(scrapy.Spider):
+
+        next_page = response.css('span.pages__item__active').get()
+        if next_page is not None:
+            yield response.follow(next_page, callback=self.parse)
```
Ostateczenie zmienna `next_page` , którą przekazujesz do metody `.follow()` to musi być pełen link kolejnej strony np. `https://ogloszenia.trojmiasto.pl/nieruchomosci/dom/gdynia/?strona=1` , gdzie numer strony będzie się zmieniał w zależności od strony, ale ta informacja będzie pobierana z elementu HTML <a> z jego atrybutu HTLM, tak jak w komentarzu powyżej.

bagaw1 committed:
add fix to elements interation

bagaw1 committed:
add fix

anetapec committed:
Poki co cena, url i powierzchnia

anetapec committed:
test

bagaw1 committed:
add next page scrapping

anetapec committed:
raed.me, pipelines, wyliczona cena za metr

anetapec committed:
Create virtual environment:
  python3 -m venv venv

Activate VENV
  source venv/bin/activate

Install scrapy
  pip install scrapy

Runing scrapy and writing data to file tricity.csv
  scrapy crawl tricity -o tricity.csv

anetapec committed:
jeszcze lczba pokoi

anetapec committed:
na ta popatrz

anetapec commented:
Jak ominąć problem gdy program się wysypuje podczas natrafienia ogłoszenia bez danych które ściągamy

anetapec reviewed (commented):
próba przywrócenia

anetapec committed:
zeskrobalo ponad 6780 pozycji , nie uwzglednilo filtrow ktore narzucilismy na poczatku

anetapec committed:
proba hashowania

anetapec commented:
czy to o to chodzi z tym ID ? program nie działa ..ale powiedz czy to miałeś na mysli ....

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -1,20 +1,50 @@
+                house_item['url'] = house.css('a.list__item__content__title__name.link').attrib['href']
+                house_item['numbers_of_rooms'] = re.sub("[^.0-9]", "", house.css('li.list__item__details__icons__element.details--icons--element--l_pokoi').get())
+                house_item['area'] = re.sub("[^.0-9]", "",house.css('p.list__item__details__icons__element__desc::text').get())
+                #house_item['price_per_meter'] = float(round((house_item['price']) / float(house_item['area'])), 2)
```
To będziemy wyliczać w Pipelines (dalszym procesowaniu danych). A nie w tym miejscu.

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -1,20 +1,50 @@
 
-            yield house_item
\ No newline at end of file
+        for house_item in HouseItem():
```
To też będziemy wyliczać w Pipelines (dalszym procesowaniu danych). Nie w momencie pobierania danych.
https://docs.scrapy.org/en/latest/topics/item-pipeline.html. Usuń narazie ten kawałek kodu z tego brancha. Stwórz sobie inny branch. Tego będziemy mergować.

bagaw1 commented on tricity/tricity/spiders/trojmiasto.py:
```diff
@@ -1,20 +1,50 @@
+                #house_item['price_per_meter'] = float(round((house_item['price']) / float(house_item['area'])), 2)
+                house_item['price_per_meter'] = float(house_item['price']) / float(house_item['area'])
+            
+            except: 
```
Tak o to chodziło. To jest jeden ze sposobów jak można to rozwiązać.

anetapec committed:
ostateczny

anetapec reviewed (commented):
ostateczny etap pobierania danych ze strony .

anetapec commented:
etap pierwszy

anetapec commented:
.
